offchain/dispatcher/src/drivers/context.rs

```diff
@@ -62,98 +67,122 @@ impl Context {
         input: &Input,
         broker: &impl BrokerSend,
     ) -> Result<(), BrokerFacadeError> {
-        broker.enqueue_input(self.inputs_sent_count, input).await?;
+        let input_block = input.block_added.number.as_u64();
+        self.finish_epoch_if_needed(input_block, broker).await?;
+
+        broker.enqueue_input(self.inputs_sent, input).await?;
+
         self.metrics
             .advance_inputs_sent
             .get_or_create(&self.dapp_metadata)
             .inc();
-        self.inputs_sent_count += 1;
-        self.last_event_is_finish_epoch = false;
+
+        self.inputs_sent += 1;
+        self.last_input_epoch =
+            Some(self.calculate_epoch(input.block_added.number.as_u64()));
+
         Ok(())
     }
 }
 
 impl Context {
-    fn calculate_epoch(&self, timestamp: u64) -> u64 {
-        assert!(timestamp >= self.genesis_timestamp);
-        (timestamp - self.genesis_timestamp) / self.epoch_length
-    }
-
-    // This logic works because we call this function with `event_timestamp` being equal to the
-    // timestamp of each individual input, rather than just the latest from the blockchain.
-    fn should_finish_epoch(&self, event_timestamp: u64) -> bool {
-        if self.inputs_sent_count == 0 || self.last_event_is_finish_epoch {
-            false
-        } else {
-            let current_epoch = self.calculate_epoch(self.last_timestamp);
-            let event_epoch = self.calculate_epoch(event_timestamp);
-            event_epoch > current_epoch
+    fn calculate_epoch(&self, block: u64) -> u64 {
+        assert!(block >= self.genesis_block);
+        (block - self.genesis_block) / self.epoch_length
+    }
+
+    fn should_finish_epoch(&self, epoch: u64) -> bool {
+        match (self.last_input_epoch, self.last_finished_epoch) {
```

Comment:
Shouldn't this assert be elsewhere? eg.: When writing this values?

Reply:
Also created this assertion in `finish_epoch`:
```
assert!(self.last_input_epoch.is_some());
self.last_finished_epoch = self.last_input_epoch;
```